bee-ledger/src/workers/snapshot/import.rs

```diff
@@ -0,0 +1,317 @@
+        storage::{self, apply_balance_diffs, apply_milestone, create_output, rollback_milestone, StorageBackend},
+    },
+};
+
+use bee_common::packable::{Packable, Read};
+use bee_message::{
+    milestone::MilestoneIndex,
+    output::{self, Output, OutputId},
+    payload::Payload,
+    MessageId,
+};
+use bee_storage::access::{Insert, Truncate};
+use bee_tangle::solid_entry_point::SolidEntryPoint;
+
+use chrono::{offset::TimeZone, Utc};
+use log::{info, warn};
+
+use std::{
+    collections::HashMap,
+    fs::{File, OpenOptions},
+    io::BufReader,
+    path::Path,
+};
+
+fn snapshot_reader(path: &Path) -> Result<BufReader<File>, Error> {
+    Ok(BufReader::new(
+        OpenOptions::new()
+            .read(true)
+            .open(path)
+            .map_err(|e| Error::Snapshot(SnapshotError::Io(e)))?,
+    ))
+}
+
+async fn import_solid_entry_points<R: Read, B: StorageBackend>(
+    reader: &mut R,
+    storage: &B,
+    sep_count: u64,
+    index: MilestoneIndex,
+) -> Result<(), Error> {
+    Truncate::<SolidEntryPoint, MilestoneIndex>::truncate(storage)
+        .await
+        .map_err(|e| Error::Storage(Box::new(e)))?;
+    for _ in 0..sep_count {
+        Insert::<SolidEntryPoint, MilestoneIndex>::insert(&*storage, &SolidEntryPoint::unpack(reader)?, &index)
+            .await
+            .map_err(|e| Error::Storage(Box::new(e)))?;
+    }
+
+    Ok(())
+}
+
+async fn import_outputs<R: Read, B: StorageBackend>(
+    reader: &mut R,
+    storage: &B,
+    output_count: u64,
+) -> Result<(), Error> {
+    let mut balance_diffs = BalanceDiffs::new();
+
+    for _ in 0..output_count {
+        let message_id = MessageId::unpack(reader)?;
+        let output_id = OutputId::unpack(reader)?;
+        let output = Output::unpack(reader)?;
+        let created_output = CreatedOutput::new(message_id, output);
+
+        create_output(&*storage, &output_id, &created_output).await?;
+        balance_diffs.output_add(created_output.inner())?;
+    }
+
+    apply_balance_diffs(&*storage, &balance_diffs).await?;
+
+    Ok(())
+}
+
+async fn import_milestone_diffs<R: Read, B: StorageBackend>(
+    reader: &mut R,
+    storage: &B,
+    milestone_diff_count: u64,
+) -> Result<(), Error> {
+    for _ in 0..milestone_diff_count {
+        let diff = MilestoneDiff::unpack(reader)?;
+        let index = diff.milestone().essence().index();
+        // Unwrap is fine because ledger index was inserted just before.
+        let ledger_index = *storage::fetch_ledger_index(&*storage).await?.unwrap();
+        let mut balance_diffs = BalanceDiffs::new();
+        let mut consumed = HashMap::new();
+
+        for (_, output) in diff.created().iter() {
+            balance_diffs.output_add(output.inner())?;
+        }
+
+        for (output_id, (created_output, consumed_output)) in diff.consumed().iter() {
+            balance_diffs.output_sub(created_output.inner())?;
+            consumed.insert(*output_id, (created_output.clone(), consumed_output.clone()));
```

Comment:
It doesn't look like `consumed` is mutated after this line. Consider a `map` + `collect` instead? It'll likely result in fewer allocations too.